src/main/java/org/eclipsefoundation/adopters/service/impl/DefaultAdopterService.java

```diff
@@ -0,0 +1,213 @@
+import java.nio.file.WatchService;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.stream.Collectors;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import javax.enterprise.context.ApplicationScoped;
+
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+import org.eclipsefoundation.adopters.model.AdoptedProject;
+import org.eclipsefoundation.adopters.model.Adopter;
+import org.eclipsefoundation.adopters.model.AdopterList;
+import org.eclipsefoundation.adopters.model.Project;
+import org.eclipsefoundation.adopters.service.AdopterService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+
+import io.quarkus.runtime.Startup;
+
+/**
+ * Retrieves the adopters information from the filesystem on service start, and
+ * provides copies to requesting callers.
+ * 
+ * @author Martin Lowe
+ *
+ */
+@Startup
+@ApplicationScoped
+public class DefaultAdopterService implements AdopterService {
+	private static final Logger LOGGER = LoggerFactory.getLogger(DefaultAdopterService.class);
+
+	@ConfigProperty(name = "eclipse.adopters.dirpath", defaultValue = "/tmp/quarkus")
+	String adoptersLocation;
+	@ConfigProperty(name = "eclipse.adopters.filename", defaultValue = "adopters.json")
+	String adoptersName;
+	@ConfigProperty(name = "eclipse.adopters.retry.count", defaultValue = "3")
+	int retryCount;
+	@ConfigProperty(name = "eclipse.adopters.retry.timeoutInMS", defaultValue = "1000")
+	long retryTimeout;
+
+	private List<Adopter> adopters;
+	private WatchService watcher;
+
+	private ReentrantLock lock;
+	private boolean running;
+
+	@PostConstruct
+	public void init() throws IOException {
+		this.lock = new ReentrantLock();
+		this.adopters = new ArrayList<>();
+		Path adoptersDir = Paths.get(adoptersLocation);
+		Path adoptersFile = Paths.get(adoptersLocation, adoptersName);
+		if (Files.exists(adoptersDir)) {
+			// read in the initial file if it exists
+			if (Files.exists(adoptersFile)) {
+				LOGGER.debug("Found an adopters file at path {}, reading in", adoptersFile);
+				readInFile(adoptersFile);
+			}
+
+			watcher = FileSystems.getDefault().newWatchService();
+			// register the adopter location to watch
+			LOGGER.debug("Registering file watcher on directory {}", adoptersDir);
+			adoptersDir.register(watcher, ENTRY_CREATE, ENTRY_DELETE, ENTRY_MODIFY);
+
+			// start file watching on a new thread
+			new Thread(this::update).start();
+		} else {
+			throw new IllegalArgumentException();
+		}
+	}
+
+	@PreDestroy
+	public void shutdown() {
+		this.running = false;
+	}
+
+	/**
+	 * Update loop, will loop until shutdown is triggered or the service is stopped.
+	 * 
+	 * @throws IOException
+	 */
+	private void update() {
+		// soft lock this to a single thread
+		if (!this.lock.isLocked()) {
+			this.lock.lock();
+			try {
+				// set the state to running at the start
+				this.running = true;
```

Comment:
Either use a AtomicBoolean or a volatile boolean to force the "happens-before" ordering or write/read operation.

However, I would just use a `while(true)` loop and rely on interrupt to stop the thread. It's less code, and it relies on JVM behavior. In order to do that, you keep a reference to the watcher thread in a field and in `PreDestroy` you call watcherThread.interrupt. Then, in `update`, when you catch an `InterruptedException`, you just return, effectively completing the thread.